pants_jupyter_plugin/plugin.py

```diff
@@ -0,0 +1,403 @@
+    ) -> int:
+      p = await asyncio.create_subprocess_shell(
+        cmd,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.STDOUT
+      )
+
+      while True:
+        line = await p.stdout.readline()
+        if not line: break
+        display(line.decode())
+
+      try:
+        return_code = await p.wait()
+      finally:
+        is_complete.set()
+
+      return return_code
+
+    def run_async(executor, spinner):
+      nest_asyncio.apply()
+      loop = asyncio.get_event_loop()
+      finished, unfinished = loop.run_until_complete(
+        asyncio.wait(
+          [executor, spinner],
+          return_when=asyncio.ALL_COMPLETED
+        )
+      )
+      assert len(finished) == 2, f'unexpected async execution results: finished={finished}'
+      assert not unfinished, f'unexpected async execution results: unfinished={unfinished}'
+
+      results = [r for r in [task.result() for task in finished] if r is not None]
+      assert len(results) == 1, f'unexpected results: {results}'
+      return_code = results[0]
+
+      if return_code != 0:
+        raise self.SubprocessFailure(
+          f'command `{cmd}` failed with exit code {return_code}',
+          return_code=return_code
+        )
+
+    with self._accordion_widget(title, collapsed=False) as (expand, collapse, set_output_glyph):
+      self._display_line(f'$ {cmd}\n')
+      is_complete = asyncio.Event()
+
+      try:
+        run_async(
+          async_exec(self._display_line, cmd, title, is_complete),
+          spin_driver(set_output_glyph, is_complete)
+        )
+        resulting_binary = self._extract_resulting_binary(work_dir, extension)
+        self._display_line(f'\nSuccessfully built {resulting_binary}')
+
+        set_output_glyph(SUCCESS_GLYPH)
+        collapse()
+        return resulting_binary
+      except self.SubprocessFailure:
+        try:
+          set_output_glyph(FAIL_GLYPH)
+          expand()
+          self._display_line('\n\n')
+        finally:
+          raise
+
+  def _run_pex(self, requirements: str) -> pathlib.PosixPath:
+    """Runs pex with widget UI display."""
+    with temporary_dir(cleanup=False) as tmp_dir:
+      tmp_path = pathlib.PosixPath(tmp_dir)
+      output_pex = tmp_path.joinpath('requirements.pex')
+      title = f'[Resolve] {requirements}'
+      safe_requirements = ' '.join(shlex.quote(r) for r in shlex.split(requirements))
+      # TODO: Add support for toggling `--no-pypi` and find-links/index configs.
+      cmd = f'pex -vv -o "{output_pex}" {safe_requirements}'
+      return self._stream_binary_build_with_output(cmd, title, tmp_path, extension='pex')
+
+  def _run_pants(
+    self,
+    pants_repo: pathlib.PosixPath,
+    pants_target: str,
+    extension: str
+  ) -> pathlib.PosixPath:
+    """Runs pants with widget UI display."""
+
+    # Version check for pants v1 vs v2 flags/behavior.
+    is_pants_v1 = pants_repo.joinpath('pants.ini').exists()
+    if is_pants_v1:
```

Comment:
Technically `pants.toml` became an option at `1.26.x`. Not sure you want to pay the cost, but a `./pants --version` check and string comparison for `2.*` would be a little bit more bulletproof (`./pants binary` and `package` both existed at 2.0, with `binary` deprecated).

Reply:
I suppose we could run `./pants --version` when `%pants_repo` is set to make that a one-time cost. in a cold repo, this might result in a longer than expected bootstrap and potential failures tho - but probably worth it for multi-version support (as we're unfortunately stuck on v1 for probably another 6-12+ months).

Reply:
Not a blocker probably. I can file an issue about it, and we can focus the immediate efforts on getting it in at all with some test coverage.